prow/tide/contextregister.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tide
+
+import (
+	"github.com/shurcooL/githubql"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/config"
+)
+
+type contextChecker interface {
+	// isOptional tells whether a context is optional.
+	isOptional(Context) bool
+	// missingRequiredContexts tells if required contexts are missing from the list of contexts provided.
+	missingRequiredContexts([]Context) []Context
+}
+
+// newExpectedContext creates a Context with Expected state.
+func newExpectedContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// contextRegister implements contextChecker and allow registering of required and optional contexts.
+type contextRegister struct {
```

Comment:
I tend to agree with you, but that would create circular dependency and I would have to move the Context struct out as an example. Plus I would have to hard code the tide status in there. I am trying to keep this as the minimum amount of change. However, I don't think that the context checker interface is something that we need. The implementation is very general and straightforward. What do you think ?

Reply:
What is the circular dependency? The `contextRegister` is basically copying the fields from `TideContextPolicy` and defining new methods so I'd really prefer to condense these into one struct and just configure the packages and imports properly to avoid a circular dependency.

The tide status shouldn't need to be hard coded? It can be passed down from the `tide` package.

Reply:
The circular deps is on the Context object defined in prow/tide/tide.go

Reply:
Done